File: native-plugins/capacitor-game-services/ios/Sources/CapacitorGameServices/CapacitorGameServicesPlugin.swift
import Capacitor
import Foundation

@objc(CapacitorGameServicesPlugin)
public class CapacitorGameServicesPlugin: CAPPlugin, CAPBridgedPlugin {
    private let storageLoadProtocol = "mpgd.storage.load.v1"
    public let identifier = "CapacitorGameServicesPlugin"
    public let jsName = "CapacitorGameServices"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "request", returnType: CAPPluginReturnPromise)
    ]
    private lazy var localStorage = LocalJsonStorage(
        backend: MigratingFileLocalJsonStorageBackend(
            backend: FileLocalJsonStorageBackend(),
            legacyDefaults: .standard
        )
    )

    @objc func request(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), !id.isEmpty else {
            call.resolve(errorResponse(id: "", code: "INVALID_BRIDGE_REQUEST", message: "Bridge request ID is required."))
            return
        }

        guard let method = call.getString("method"), !method.isEmpty else {
            call.resolve(errorResponse(id: id, code: "INVALID_BRIDGE_REQUEST", message: "Bridge method is required."))
            return
        }

        switch method {
        case "runtime.getCapabilities":
            call.resolve(okResponse(id: id, data: capabilities()))
        case "identity.getPlayer":
            call.resolve(okResponse(id: id, data: player()))
        case "identity.getSession":
            call.resolve(okResponse(id: id, data: [
                "identityLevel": "platform-anonymous",
                "playerId": "ios-local-player",
                "trustLevel": "local"
            ]))
        case "identity.requestUpgrade":
            call.resolve(okResponse(id: id, data: [
                "status": "unavailable",
                "reloadExpected": false
            ]))
        case "presentation.getLaunchIntent":
            call.resolve(okResponse(id: id, data: ["entry": "home"]))
        case "presentation.requestGameSurface":
            call.resolve(okResponse(id: id, data: "already-fullscreen"))
        case "share.share":
            call.resolve(okResponse(id: id, data: ["status": "unavailable"]))
        case "share.readInboundShare":
            call.resolve(okResponse(id: id, data: NSNull()))
        case "notifications.getStatus":
            call.resolve(okResponse(id: id, data: "configuration-required"))
        case "notifications.requestSubscription":
            call.resolve(okResponse(id: id, data: "unavailable"))
        case "commerce.getProducts", "commerce.purchase", "commerce.restore", "commerce.getEntitlements":
            call.resolve(errorResponse(id: id, code: "NATIVE_IAP_UNAVAILABLE", message: "No native store provider is installed."))
        case "ads.preload", "ads.showRewarded", "ads.showInterstitial", "ads.mountBanner", "ads.unmountBanner":
            call.resolve(errorResponse(id: id, code: "NATIVE_ADS_UNAVAILABLE", message: "No native ads provider is installed."))
        case "leaderboard.submitScore", "leaderboard.open":
            call.resolve(errorResponse(id: id, code: "NATIVE_LEADERBOARD_UNAVAILABLE", message: "No native leaderboard provider is installed."))
        case "storage.load":
            loadStorage(call, id: id)
        case "storage.save":
            saveStorage(call, id: id)
        default:
            call.resolve(errorResponse(id: id, code: "UNSUPPORTED_METHOD", message: "Unsupported bridge method: \(method)"))
        }
    }

    private func okResponse(id: String, data: Any) -> [String: Any] {
        return [
            "id": id,
            "ok": true,
            "data": data
        ]
    }

    private func errorResponse(
        id: String,
        code: String,
        message: String,
        retryable: Bool = false
    ) -> [String: Any] {
        return [
            "id": id,
            "ok": false,
            "error": [
                "code": code,
                "message": message,
                "retryable": retryable
            ]
        ]
    }

    private func loadStorage(_ call: CAPPluginCall, id: String) {
        guard let key = storageKey(call) else {
            call.resolve(errorResponse(
                id: id,
                code: "INVALID_STORAGE_KEY",
                message: "Storage key must be a string."
            ))
            return
        }

        do {
            guard let serializedValue = try localStorage.load(key: key) else {
                call.resolve(okResponse(id: id, data: [
                    "__mpgdBridgeProtocol": storageLoadProtocol,
                    "found": false
                ]))
                return
            }

            let value = try JSONSerialization.jsonObject(
                with: Data(serializedValue.utf8),
                options: [.fragmentsAllowed]
            )
            call.resolve(okResponse(id: id, data: [
                "__mpgdBridgeProtocol": storageLoadProtocol,
                "found": true,
                "value": value
            ]))
        } catch let error as LocalJsonStorageError {
            call.resolve(errorResponse(
                id: id,
                code: error.bridgeCode,
                message: error.bridgeMessage,
                retryable: error.retryable
            ))
        } catch {
            call.resolve(errorResponse(
                id: id,
                code: "NATIVE_STORAGE_LOAD_FAILED",
                message: "Native storage contained an invalid JSON value."
            ))
        }
    }

    private func saveStorage(_ call: CAPPluginCall, id: String) {
        guard let payload = call.getObject("payload"),
              let key = payload["key"] as? String else {
            call.resolve(errorResponse(
                id: id,
                code: "INVALID_STORAGE_KEY",
                message: "Storage key must be a string."
            ))
            return
        }

        guard payload.keys.contains("value"), let value = payload["value"] else {
            call.resolve(errorResponse(
                id: id,
                code: "INVALID_STORAGE_VALUE",
                message: "Storage value must be JSON serializable."
            ))
            return
        }

        let serializedValue: String

        do {
            let data = try JSONSerialization.data(
                withJSONObject: value,
                options: [.fragmentsAllowed]
            )

            guard let candidate = String(data: data, encoding: .utf8) else {
                throw EncodingError.invalidValue(
                    value,
                    EncodingError.Context(
                        codingPath: [],
                        debugDescription: "JSON serialization did not produce UTF-8."
                    )
                )
            }

            serializedValue = candidate
        } catch {
            call.resolve(errorResponse(
                id: id,
                code: "INVALID_STORAGE_VALUE",
                message: "Storage value must be JSON serializable."
            ))
            return
        }

        do {
            try localStorage.save(key: key, serializedValue: serializedValue)
            call.resolve(okResponse(id: id, data: ["saved": true]))
        } catch let error as LocalJsonStorageError {
            call.resolve(errorResponse(
                id: id,
                code: error.bridgeCode,
                message: error.bridgeMessage,
                retryable: error.retryable
            ))
        } catch {
            call.resolve(errorResponse(
                id: id,
                code: "NATIVE_STORAGE_SAVE_FAILED",
                message: "Native storage could not be saved.",
                retryable: true
            ))
        }
    }

    private func storageKey(_ call: CAPPluginCall) -> String? {
        return call.getObject("payload")?["key"] as? String
    }

    private func capabilities() -> [String: Any] {
        return [
            "nativeIap": false,
            "nativeAds": false,
            "rewardedAds": false,
            "interstitialAds": false,
            "bannerAds": false,
            "nativeLeaderboard": false,
            "remoteLeaderboard": false,
            "achievements": false,
            "cloudSave": false,
            "socialShare": false,
            "haptics": false,
            "localizedContent": true
        ]
    }

    private func player() -> [String: Any] {
        return [
            "playerId": "ios-local-player",
            "displayName": "iOS Local Player"
        ]
    }

}

File: native-plugins/capacitor-game-services/android/src/main/java/dev/mpgd/capacitor/CapacitorGameServicesPlugin.java
package dev.mpgd.capacitor;

import android.content.Context;
import android.content.SharedPreferences;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import org.json.JSONObject;

@CapacitorPlugin(name = "CapacitorGameServices")
public class CapacitorGameServicesPlugin extends Plugin {
    private static final String STORAGE_PREFERENCES =
        "dev.mpgd.capacitor.gameservices.storage";
    private LocalJsonStorage localStorage;

    @PluginMethod
    public void request(PluginCall call) {
        String id = call.getString("id");
        String method = call.getString("method");

        if (id == null || id.isEmpty()) {
            id = "android-native-mock";
        }

        if (method == null || method.isEmpty()) {
            call.resolve(errorResponse(id, "INVALID_BRIDGE_REQUEST", "Bridge method is required."));
            return;
        }

        switch (method) {
            case "runtime.getCapabilities":
                call.resolve(okResponse(id, capabilities()));
                return;
            case "identity.getPlayer":
                call.resolve(okResponse(id, player()));
                return;
            case "identity.getSession":
                call.resolve(okResponse(id, new JSObject()
                    .put("identityLevel", "platform-anonymous")
                    .put("playerId", "android-local-player")
                    .put("trustLevel", "local")));
                return;
            case "identity.requestUpgrade":
                call.resolve(okResponse(id, new JSObject()
                    .put("status", "unavailable")
                    .put("reloadExpected", false)));
                return;
            case "presentation.getLaunchIntent":
                call.resolve(okResponse(id, new JSObject().put("entry", "home")));
                return;
            case "presentation.requestGameSurface":
                call.resolve(okResponse(id, "already-fullscreen"));
                return;
            case "share.share":
                call.resolve(okResponse(id, new JSObject().put("status", "unavailable")));
                return;
            case "share.readInboundShare":
                call.resolve(okResponse(id, JSONObject.NULL));
                return;
            case "notifications.getStatus":
                call.resolve(okResponse(id, "configuration-required"));
                return;
            case "notifications.requestSubscription":
                call.resolve(okResponse(id, "unavailable"));
                return;
            case "commerce.getProducts":
                call.resolve(okResponse(id, new Object[] { product() }));
                return;
            case "commerce.purchase":
                call.resolve(okResponse(id, new JSObject()
                    .put("status", "completed")
                    .put("transactionId", "android-mock-" + id)
                    .put("entitlementIds", new String[] { "COINS_100" })));
                return;
            case "commerce.restore":
                call.resolve(okResponse(id, new JSObject().put("restoredEntitlements", new Object[] {})));
                return;
            case "commerce.getEntitlements":
                call.resolve(okResponse(id, new Object[] {}));
                return;
            case "ads.preload":
                call.resolve(okResponse(id, new JSObject()));
                return;
            case "ads.showRewarded":
                call.resolve(okResponse(id, new JSObject()
                    .put("status", "completed")
                    .put("rewardGranted", true)
                    .put("ledgerEntryId", "android-reward-" + id)));
                return;
            case "ads.showInterstitial":
                call.resolve(okResponse(id, new JSObject().put("status", "shown")));
                return;
            case "leaderboard.submitScore":
                call.resolve(okResponse(id, new JSObject().put("submitted", true)));
                return;
            case "leaderboard.open":
                call.resolve(okResponse(id, new JSObject().put("opened", true)));
                return;
            case "storage.load":
                loadStorage(call, id);
                return;
            case "storage.save":
                saveStorage(call, id);
                return;
            default:
                call.resolve(errorResponse(id, "UNSUPPORTED_METHOD", "Unsupported bridge method: " + method));
        }
    }

    private JSObject okResponse(String id, Object data) {
        return new JSObject()
            .put("id", id)
            .put("ok", true)
            .put("data", data);
    }

    private JSObject errorResponse(String id, String code, String message) {
        return errorResponse(id, code, message, false);
    }

    private JSObject errorResponse(String id, String code, String message, boolean retryable) {
        return new JSObject()
            .put("id", id)
            .put("ok", false)
            .put("error", new JSObject()
                .put("code", code)
                .put("message", message)
                .put("retryable", retryable));
    }

    private void loadStorage(PluginCall call, String id) {
        String key = storageKey(call);

        if (key == null) {
            call.resolve(errorResponse(
                id,
                "INVALID_STORAGE_KEY",
                "Storage key must be a string."
            ));
            return;
        }

        try {
            String serializedValue = localStorage().load(key);
            Object value;

            if (serializedValue == null) {
                value = JSONObject.NULL;
            } else {
                JSONObject wrapper = new JSONObject(serializedValue);

                if (!wrapper.has("value")) {
                    throw new IllegalStateException("Stored JSON wrapper is missing its value.");
                }

                value = wrapper.opt("value");
            }

            call.resolve(okResponse(id, value));
        } catch (LocalJsonStorage.StorageException error) {
            call.resolve(errorResponse(
                id,
                error.getCode(),
                error.getMessage(),
                error.isRetryable()
            ));
        } catch (Exception error) {
            call.resolve(errorResponse(
                id,
                "NATIVE_STORAGE_LOAD_FAILED",
                "Native storage contained an invalid JSON value."
            ));
        }
    }

    private void saveStorage(PluginCall call, String id) {
        JSObject payload = call.getObject("payload");
        String key = storageKey(payload);

        if (key == null) {
            call.resolve(errorResponse(
                id,
                "INVALID_STORAGE_KEY",
                "Storage key must be a string."
            ));
            return;
        }

        if (payload == null || !payload.has("value")) {
            call.resolve(errorResponse(
                id,
                "INVALID_STORAGE_VALUE",
                "Storage value must be JSON serializable."
            ));
            return;
        }

        final String serializedValue;

        try {
            serializedValue = new JSONObject()
                .put("value", payload.opt("value"))
                .toString();
        } catch (Exception error) {
            call.resolve(errorResponse(
                id,
                "INVALID_STORAGE_VALUE",
                "Storage value must be JSON serializable."
            ));
            return;
        }

        try {
            localStorage().save(key, serializedValue);
            call.resolve(okResponse(id, new JSObject().put("saved", true)));
        } catch (LocalJsonStorage.StorageException error) {
            call.resolve(errorResponse(
                id,
                error.getCode(),
                error.getMessage(),
                error.isRetryable()
            ));
        }
    }

    private String storageKey(PluginCall call) {
        return storageKey(call.getObject("payload"));
    }

    private String storageKey(JSObject payload) {
        return payload == null ? null : payload.getString("key");
    }

    private synchronized LocalJsonStorage localStorage() {
        if (localStorage != null) {
            return localStorage;
        }

        SharedPreferences preferences = getContext().getSharedPreferences(
            STORAGE_PREFERENCES,
            Context.MODE_PRIVATE
        );
        localStorage = new LocalJsonStorage(new LocalJsonStorage.Backend() {
            @Override
            public String get(String key) {
                return preferences.getString(key, null);
            }

            @Override
            public boolean put(String key, String value) {
                boolean hadPreviousValue = preferences.contains(key);
                String previousValue = hadPreviousValue ? preferences.getString(key, null) : null;
                boolean committed = preferences.edit().putString(key, value).commit();

                if (!committed) {
                    SharedPreferences.Editor rollback = preferences.edit();

                    if (hadPreviousValue) {
                        rollback.putString(key, previousValue);
                    } else {
                        rollback.remove(key);
                    }

                    rollback.commit();
                }

                return committed;
            }
        });
        return localStorage;
    }

    private JSObject capabilities() {
        return new JSObject()
            .put("nativeIap", true)
            .put("nativeAds", true)
            .put("rewardedAds", true)
            .put("interstitialAds", true)
            .put("nativeLeaderboard", true)
            .put("achievements", false)
            .put("cloudSave", false)
            .put("socialShare", false)
            .put("haptics", true)
            .put("localizedContent", true);
    }

    private JSObject player() {
        return new JSObject()
            .put("playerId", "android-local-player")
            .put("displayName", "Android Local Player");
    }

    private JSObject product() {
        return new JSObject()
            .put("id", "COINS_100")
            .put("type", "consumable")
            .put("title", "100 Coins")
            .put("description", "Adds 100 demo coins.")
            .put("price", new JSObject()
                .put("formatted", "$0.99")
                .put("currencyCode", "USD"));
    }
}
